### ISSUE TITLE
runtime: hugepage-aware heap

### ISSUE THREAD

**Issue by mknyszek** [golang/go#59960]
## Background

[Transparent huge pages (THP) admin guide](https://www.kernel.org/doc/html/latest/admin-guide/mm/transhuge.html).
[Go scavenging policy](https://go.googlesource.com/proposal/+/master/design/30333-smarter-scavenging.md#which-memory-should-we-scavenge). (Implementation details are out-of-date, but linked policy is relevant.)
[THP flag behavior]().

## Motivation

Currently, Go's hugepage-related policies [do not play well together](https://github.com/golang/go/issues/55328) and have bit-rotted. The result is that the memory regions the Go runtime chooses to mark as `MADV_NOHUGEPAGE` and `MADV_HUGEPAGE` are somewhat haphazard, resulting in memory overuse for small heaps. The memory overuse is upwards of 40% memory overhead in some cases. Turning off huge pages entirely fixes the problem, but leaves CPU performance on the table. This policy also means large heaps might have dense sections that are erroneously mapped as `MADV_NOHUGEPAGE`, costing up to 1% throughput.

The goal of this work is to eliminate this overhead for small heaps while improving huge page utilization for large heaps.

## Design

Full design doc. (Shortly)

TL;DR: Determine which parts of the heap should be backed by huge pages based on near-future density of page allocator chunks. This density heuristic is then also used to delay the background scavenger: only chunks that have been sparse for at least one GC cycle are eligible for scavenging. The scavenger also breaks up hugepages.

**Comment by mknyszek** [comment]
Landed in https://go.dev/cl/436395.
